Setup /editprofile endpoint

JrodanDiaz opened the pull request:
## Summary

- Created an `editUser()` func in data/user.go that interacts with DB
- Modified Enzo's `AuthenticateRoute` middleware so that it's request passes the verified token object to verified routes via context (The purpose of this is to get the Subject from the Token Claims, which is then passed to `editUser()` so that it knows which row to update)
- Created `editProfile()` HTTP handler in handlers/user.go
- Tested that everything works with Postman
- NOTE: I could not fix the Prettier errors that wouldn't let me commit on Monday, so i ran `git commit --no-verify`. Let me know if this PR will break everything 😅

lesterfernandez commented on server/handlers/auth.go:
```diff
@@ -1,12 +1,17 @@
 
+type ctxKey string
+
+var ContextKey ctxKey
```
After setting this variable in `AuthenticateRoute`, it never changes. Could we turn this into a `const` and set its value here instead?

lesterfernandez commented on server/handlers/user.go:
```diff
@@ -103,3 +103,34 @@ func (s *Server) loginImplicitly(w http.ResponseWriter, res *http.Request) {
+	subject, tokenErr := verifiedToken.Claims.GetSubject()
+	if tokenErr != nil {
+		//not sure what error msg is best here
+		respondWithError(w, "Invalid Token", http.StatusBadRequest)
```
Since it's an invalid authorization token, maybe we should send back a status code of unauthorized?

lesterfernandez commented on server/handlers/user.go:
```diff
@@ -103,3 +103,34 @@ func (s *Server) loginImplicitly(w http.ResponseWriter, res *http.Request) {
+	user := data.UserProfile{}
+	decodeErr := json.NewDecoder(req.Body).Decode(&user)
+	if decodeErr != nil {
+		respondWithError(w, "Internal Server Error", http.StatusInternalServerError)
```
If the request's body didn't match the struct we expected it's not the server's fault, it's the client's fault. Maybe we can change this status code to bad request?

lesterfernandez commented on server/handlers/user.go:
```diff
@@ -103,3 +103,34 @@ func (s *Server) loginImplicitly(w http.ResponseWriter, res *http.Request) {
+	}
+	updatedUser, updateErr := s.User.EditUser(&user, subject)
+	if updateErr != nil {
+		respondWithError(w, "Internal Server Error: Bad DB Query", http.StatusInternalServerError)
```
nitpick: IRL it's not a good idea to tell the client what happened with our database. Since they could use that information to their advantage if trying to attack us. But yes, in this case it is an internal server error, we just don't want to give out TMI lol. Just the status code and a generic error message suffice.

lesterfernandez reviewed (approved):
I left some nitpicks. They don't matter too much but I just want to encourage best practices. Nice code tho!

lesterfernandez commented on server/data/user.go:
```diff
@@ -142,3 +144,21 @@ func isValidLogin(username string, password string) (bool, error) {
+		return &UserProfile{}, err
+	}
+
+	updatedUser, err := getUser(username)
```
Ooh, I like this suggestion. It would make the endpoint virtually twice as fast.

JrodanDiaz commented:
- Updated the SQL query to return the updated UserProfile
- Changed `ContextKey` to a const, initialized it outside the function
- Updated the error messages and status codes to make more sense

enzofalone reviewed (approved):
LGTM!
